データベース設計

marmite334 opened the pull request:
#What
データベースの設計

#Why
アプリケーション作成に必要なテーブル・カラム情報の洗い出し

#ER図の画像URL
https://gyazo.com/ca3408444e6950b1f519f808cb44191a

marmite334 committed:
データベース設計

marmite334 committed:
カラムオプションの軽微な修正

marmite334 committed:
カラムオプションの追加修正

isyt729 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| email    | string | null:false |
+| password | string | null:false |
 
-Things you may want to cover:
+### Association
+_ has_one  :identification
+_ has_many :items
 
-* Ruby version
+# identificationsテーブル
+| Column          | Type       | Options                       |
+|-----------------|------------|-------------------------------|
+| first_name      | string     | null:false                    |
+| last_name       | string     | null:false                    |
+| kana_first_name | string     | null:false                    |
+| kana_last_name  | string     | null:false                    |
+| birth_date      | integer    | null:false                    |
+| user            | references | null:false, foreign_key: true |
```
usersテーブルとidentifivcationsテーブルを分ける意味は特段ないと思われますので、一つのテーブルにまとめましょう。

isyt729 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| last_name       | string     | null:false                    |
+| kana_first_name | string     | null:false                    |
+| kana_last_name  | string     | null:false                    |
+| birth_date      | integer    | null:false                    |
```
| birth_date      | integer    | null:false                    |

生年月日を適切に保存するため、date型を用いましょう。

isyt729 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| name        | string     | null:false                    |
+| explanation | text       | null:false                    |
+| price       | integer    | null:false                    |
+| category    | integer    | null:false                    |
+| state       | integer    | null:false                    |
+| charges     | integer    | null:false                    |
+| area        | integer    | null:false                    |
+| days        | integer    | null:false                    |
```
| category    | integer    | null:false                    |
| state       | integer    | null:false                    |
| charges     | integer    | null:false                    |
| area        | integer    | null:false                    |
| days        | integer    | null:false                    |

Active_hashを用いる際は、_idをカラム名の末尾につけましょう。

isyt729 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+_ belongs_to :user
 
-* Database initialization
+# credit_cardsテーブル
+| Column        | Type       | Options                       |
+|---------------|------------|-------------------------------|
+| information   | integer    | null:false                    |
+| month         | integer    | null:false                    |
+| year          | integer    | null:false                    |
+| security_code | integer    | null:false                    |
+| item          | references | null:false, foreign_key: true |
```
商品を購入するたびにカード情報を入力するため、カード情報を管理するテーブルは必要ありません

isyt729 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+# shipping_areasテーブル
+| Column         | Type       | Options                       |
+|----------------|------------|-------------------------------|
+| postal_code    | integer    | null:false                    |
```
| postal_code    | integer    | null:false                    |

郵便番号にはハイフンを含める必要がありますので、string型としましょう。

isyt729 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+| municipalities | string     | null:false                    |
+| address        | string     | null:false                    |
+| building       | string     |                               |
+| phone_number   | integer    | null:false                    |
```
| phone_number   | integer    | null:false                    |
0から始まる数列（電話番号）を保存できるようにするため、string型としましょう。

isyt729 replied:
こちらが修正されておりませんのでご確認お願いします。

isyt729 commented on README.md:
```diff
@@ -1,24 +1,69 @@
 # README
```
購入管理テーブルがたりませんので、追加しましょう。
誰がどの商品を購入したかを管理するために必要なテーブルです。

 また、あわせてアソシエーション組み直しをしましょう。

isyt729 reviewed (commented):
レビュー依頼ありがとうございます。
7点コメントしましたので、ご確認ください。

marmite334 committed:
コードレビュー指摘箇所の修正

・usersテーブルとidentifivcationsテーブルの統合
・birth_dateカラムの型変更
・Active_hashを用いるカラム名に_idを追加
・クレジットカード情報を管理するテーブルの削除
・postal_codeカラムの型変更
・phone_numberカラムの型変更
・購入管理テーブルの作成

marmite334 commented:
# What
・usersテーブルとidentifivcationsテーブルの統合
・birth_dateカラムの型変更
・Active_hashを用いるカラム名に_idを追加
・クレジットカード情報を管理するテーブルの削除
・postal_codeカラムの型変更
・phone_numberカラムの型変更
・購入管理テーブルの作成

# Why
・テーブルの数が多いと保守性が下がるため
・日付を保存するためのdate型に変更
・Active_hashはハッシュのidを参照しているため
・クレジットカード情報を保存するテーブルを作成するのは国際的な取り組みに違反しているため
・ハイフンを含めて保存したいのでstring型へ変更
・0から始まる値を保存したいのでstring型へ変更
・購入したユーザーと出品したユーザー、商品の情報を管理しているテーブルの存在がなかったため

# ER図のURL
https://gyazo.com/885439c0e7ead232d9d16afb72d6334c

isyt729 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+|-----------------|--------|------------|
+| nickname        | string | null:false |
+| email           | string | null:false |
+| password        | string | null:false |
```
| password        | string | null:false |

Deviseのデフォルトでカラムめいが「encrypted_password」となりますので、
DB設計時も合わせておきましょう。

isyt729 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+# purchase_managementsテーブル
+| Column         | Type       | Options                       |
+|----------------|------------|-------------------------------|
+| item           | references | null:false, foreign_key: true |
+| item_user      | references | null:false, foreign_key: true |
+| purchased_user | references | null:false, foreign_key: true |
+| shipping_area  | references | null:false, foreign_key: true |
```
purchase_manegementsテーブルは商品テーブル、ユーザテーブルへの外部キーカラムを用意することで、
どのユーザが、どの商品を購入したかを管理できます。
そのため、以下のカラム構成としましょう。
・itemsテーブルへの外部キーカラム
・usresテーブルへの外部キーカラム

marmite334 replied:
レビューありがとうございました。不明点があるので質問させていただきます。
item_userを出品者、purchased_userを購入者と考え設計を行ったつもりですが、これでは管理できないとの意味でしょうか。

kuu0123 replied:
ご質問ありがとうございます。

> レビューありがとうございました。不明点があるので質問させていただきます。
> item_userを出品者、purchased_userを購入者と考え設計を行ったつもりですが、これでは管理できないとの意味でしょうか。

出品者と購入者のidを別々で管理するのであれば、そもそも購入履歴テーブルが不要という形になります。
今回のフリマアプリの使用では「購入履歴を管理する」という実装を行っていきたいため、購入管理に関するテーブルを作成しております。
その場合は、出品者はusersテーブルで管理、購入者は購入管理テーブルでuserの外部キーを取得して管理ができるため、
「item_userを出品者、purchased_userを購入者」という実装は不要になります。

上記をご確認いただいた上でもしご不明点な点がございましたら、テックコネクトにてご質問をお待ちしております。

isyt729 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+### Association
+_ belongs_to :user
+_ belongs_to :item
+_ belongs_to :shipping_area
```
_ belongs_to :shipping_area

こちらはhas_oneとしましょう。
アソシエーションの組み方に関して不明点がございましたら、カリキュラムを復習いただきますようお願いします。
https://master.tech-camp.in/curriculums/5439

isyt729 commented on README.md:
```diff
@@ -1,24 +1,60 @@
+_ belongs_to :item
+_ belongs_to :shipping_area
+
+# shipping_areasテーブル
```
shipping_areasテーブルには purchase_managementsテーブルへの外部キーカラムを追加しましょう。

isyt729 reviewed (commented):
レビュー依頼ありがとうございます。
追加で5点コメントしておりますので、ご確認ください。

marmite334 committed:
usersテーブルのカラム名変更

passwordカラムをdeviseの初期値と合わせる

marmite334 committed:
shipping_areasテーブルのカラム型変更

phone_numberのカラム型が変更できていなかったため修正

marmite334 committed:
shipping_areasテーブルの修正

商品管理テーブルへの外部キーの設定

marmite334 committed:
アソシエーションの組み直し

marmite334 commented:
# What
・usersテーブルのカラム名変更
・shipping_areasテーブルのカラム型変更
・shipping_areasテーブルの修正
・アソシエーションの組み直し

#Why
・READMEの可読性向上
・商品管理テーブルへの外部キーの設定することで間接的にusersと関係を持たせる
・phone_numberのカラム型が修正できていなかったため修正
・親子関係の組み直し

# ER図のURL
https://gyazo.com/195fbf7f1d0425ea61ccfd6223efe35b

kuu0123 commented on README.md:
```diff
@@ -1,24 +1,61 @@
+| Column             | Type   | Options    |
+|--------------------|--------|------------|
+| nickname           | string | null:false |
+| email              | string | null:false |
```
```ruby
| email              | string | null:false |
```
こちらは、deviseが用意してくれている`unique: true`をオプションとしてつけるようにしましょう。
メールアドレスとパスワードでログインする仕様なので、他のユーザーと同じメールアドレスは作成できないようにするためです。

kuu0123 commented on README.md:
```diff
@@ -1,24 +1,61 @@
+| Column              | Type       | Options                       |
+|---------------------|------------|-------------------------------|
+| postal_code         | string     | null:false                    |
+| prefectures_id      | integer    | null:false                    |
```
```ruby
	| prefectures_id      | integer    | null:false                    |
```
都道府県のカラムはactive hashを使用したいので「モデル名_id」という形でカラム名を修正しましょう。

また、都道府県に関しては、出品時に使用した「発送元の地域」と同じactive hashを使用したいと考えています。
そのため、カラム名も統一しておくといいでしょう。
どちらに合わせても問題ありませんが、「prefecture_id」に合わせた方が、「都道府県」という情報を取り扱っているとわかりやすいかもしれません。

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

marmite334 committed:
emailカラムを一意性に変更

marmite334 committed:
itemsテーブルの発送元カラムの修正

同じactive hashからデータを取得したいため

marmite334 committed:
purchase_managementsテーブルの修正

marmite334 commented:
# What
・emailカラムを一意性に変更
・itemsテーブルの発送元カラムの修正 
・purchase_managementsテーブルの修正

# Why
・同じ値を保存できないようにするため
・同じactive hashからデータを取得したいため
・アプリケーションの仕様に沿ったuser管理手法に変更

# ER図のURL
https://gyazo.com/a68dd1a9746515bc85158323be73581c

takumi-kodama commented on README.md:
```diff
@@ -1,24 +1,60 @@
+|---------------|------------|-------------------------------|
+| item          | references | null:false, foreign_key: true |
+| user          | references | null:false, foreign_key: true |
+| shipping_area | references | null:false, foreign_key: true |
```
41行目の記述は不要です。
今回purchase_managementsが親となり、shipping_areaが子となる形になっております。
そのため、アソシエーションから情報を取り出すことが可能なためです。

また、今回はフォームオブジェクトを使用して実装を行います。
一つのフォームで複数のテーブルへ保存を行えるものですが、同時ではなく、順番に処理がされます。

処理順として
購入履歴を保存→住所情報を保存という流れになります。
idは保存処理がされて初めて発番されるため、購入履歴の保存を行うタイミングでは、住所情報のidは保存されていないため付与されません。

そのため、購入履歴を保存する際に、住所情報のidがないというエラーが解決できなくなります。
住所情報保存する時には購入履歴のidが付与されているため、問題なく保存が行えます。

今回は子モデルになる`shipping_area`に外部キーを設定することで関係性を確認することが可能です。

takumi-kodama reviewed (commented):
実装お疲れ様です。
レビューのご依頼ありがとうございます。

コメントさせていただきました。ご確認をお願いいたします。

marmite334 committed:
purchase_managementsテーブルの修正

shipping_areasとのアソシエーションは親子関係で既に実現されているため。

marmite334 commented:
# What
purchase_managementsテーブルの修正

# Why
shipping_areasテーブルとのアソシエーションは親子関係で既に実現されているため。

# ER図の画像URL
https://gyazo.com/c5d6661939c13342376eb308a98ef881

masayaomoto commented on README.md:
```diff
@@ -1,24 +1,59 @@
+
+### Association
+_ belongs_to :user
+_ belongs_to :purchase_management
```
```belongs_to :purchase_management```
こちらですが、
今回の設計では「1つの商品は1つの購入情報を持っている」という設計にしたいため、
has_oneを使用して購入情報（purchase_managements）テーブルとのアソシエーションを組みましょう。

masayaomoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加で1点コメントいたしましたので、ご確認お願いします！

marmite334 committed:
アソシエーションの組み直し

marmite334 commented:
# What
アソシエーションの組直し

# Why
商品が購入できるのは1つのため、has_oneを使用する

# ER図のURL
https://gyazo.com/c5d6661939c13342376eb308a98ef881

yuevo commented:
レビュー依頼ありがとうございます！
LGTMとさせていただきます。実装お疲れ様でした！
こちらのプルリクエストは忘れずにマージしておきましょう。
引き続き実装頑張っていきましょう！
